File: app/server/narration.ts
/**
 * The text of an article, as it is meant to be heard.
 *
 * Read aloud, an article is not the page: a figure, a table and a code listing
 * are things to look at, and a voice reading their contents produces a stretch
 * of noise a listener cannot act on. They are dropped here rather than in the
 * component map, because the page keeps showing them.
 *
 * The source is the rendered HTML, not the MDX. The body is JSX by the time it
 * reaches the bucket — <Figure>, <Tableau> — and stripping tags is a smaller
 * and more honest job than parsing components: whatever the article renders as
 * is what gets read.
 */

/** What a listener hears, and the key the audio is stored under. */
export type Narration = {
  text: string
  /** Runes, not bytes — what a per-character TTS price is counted in. */
  length: number
}

const DROPPED_ELEMENTS = ['figure', 'table', 'pre', 'script', 'style']

/**
 * Elements whose end is a pause. Without them the tag-strip runs sentences
 * together — a heading straight into the paragraph below it — and the voice
 * reads across the join with no break.
 */
const BLOCK_ELEMENTS = [
  'p',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'li',
  'blockquote',
  'div',
  'section',
  'article',
  'header',
  'footer',
  'hr',
  'br',
]

export function narrationOf(article: {
  titre: string
  chapeau: string
  html: string
}): Narration {
  const text = joinParts([
    article.titre,
    article.chapeau,
    proseOf(article.html),
  ])

  return { text, length: [...text].length }
}

/**
 * The body, with everything visual removed and everything spoken kept.
 */
export function proseOf(html: string): string {
  let out = html

  for (const tag of DROPPED_ELEMENTS) {
    out = out.replace(
      new RegExp(`<${tag}\\b[^>]*>[\\s\\S]*?</${tag}>`, 'gi'),
      '\n\n',
    )
  }

  // Decorative marks carry meaning by being seen — the em dash MdxProse puts
  // before each list item is hidden from screen readers for the same reason a
  // voice must not read it as a word.
  out = out.replace(
    /<([a-z]+)\b[^>]*\baria-hidden=["']?true["']?[^>]*>[\s\S]*?<\/\1>/gi,
    ' ',
  )

  for (const tag of BLOCK_ELEMENTS) {
    out = out.replace(new RegExp(`</?${tag}\\b[^>]*>`, 'gi'), '\n\n')
  }

  out = out.replace(/<[^>]+>/g, '')

  return normalize(decodeEntities(out))
}

/**
 * A blank line between parts, which is where Split prefers to cut. Cutting
 * between paragraphs is inaudible; cutting mid-sentence is not.
 */
function joinParts(parts: string[]): string {
  return parts.map((p) => p.trim()).filter(Boolean).join('\n\n')
}

function decodeEntities(s: string): string {
  const named: Record<string, string> = {
    amp: '&',
    lt: '<',
    gt: '>',
    quot: '"',
    apos: "'",
    nbsp: ' ',
    hellip: '…',
    mdash: '—',
    ndash: '–',
    laquo: '«',
    raquo: '»',
    eacute: 'é',
    egrave: 'è',
    ecirc: 'ê',
    agrave: 'à',
    ccedil: 'ç',
    ugrave: 'ù',
    ocirc: 'ô',
    icirc: 'î',
    iuml: 'ï',
    euro: 'euros',
    // The typographic apostrophe, which MDX produces from a plain one.
    rsquo: '’',
    lsquo: '‘',
    ldquo: '“',
    rdquo: '”',
  }

  return s
    .replace(/&#(\d+);/g, (_, code: string) =>
      String.fromCodePoint(Number(code)),
    )
    .replace(/&#x([0-9a-f]+);/gi, (_, code: string) =>
      String.fromCodePoint(Number.parseInt(code, 16)),
    )
    .replace(/&([a-z]+);/gi, (whole, name: string) => named[name] ?? whole)
}

/**
 * Collapses the whitespace the tag-strip leaves behind, keeping paragraph
 * breaks: they are the only whitespace a voice reads.
 */
function normalize(s: string): string {
  return s
    .replace(/\r\n?/g, '\n')
    .replace(/[^\S\n]+/g, ' ')
    .replace(/ *\n */g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .trim()
}

File: tools/narrate/narration.go
package main

import (
	"regexp"
	"strings"
)

// narrationOf turns an article source into what a voice should say.
//
// Read aloud, an article is not the page. A figure and a table are things to
// look at: their captions and cells read as a stretch of disconnected noise a
// listener cannot act on, so they are dropped rather than narrated. What is
// left is the title, the standfirst and the prose, which is what someone would
// read out if you handed them the piece.
//
// The source is MDX, so the components are JSX tags rather than markup, and
// they are removed by name. The site does the same job from the rendered HTML
// in app/server/narration.ts — same rules, different starting point, because
// there the components have already become <figure> and <table>.
var (
	frontmatterRe = regexp.MustCompile(`(?s)\A---\r?\n(.*?)\r?\n---\r?\n`)

	// Block components, with everything they contain.
	jsxBlockRe = regexp.MustCompile(`(?s)<(Figure|Tableau)\b.*?(/>|</(Figure|Tableau)>)`)

	fencedCodeRe = regexp.MustCompile("(?s)```.*?```")
	// A markdown table: consecutive lines that start and end on a pipe.
	tableRe = regexp.MustCompile(`(?m)^\|.*\|[ \t]*$\n?`)
	imageRe = regexp.MustCompile(`!\[[^\]]*\]\([^)]*\)`)
	// A link reads as its text, not its URL.
	linkRe    = regexp.MustCompile(`\[([^\]]+)\]\([^)]*\)`)
	headingRe = regexp.MustCompile(`(?m)^#{1,6}[ \t]+`)
	quoteRe   = regexp.MustCompile(`(?m)^>[ \t]?`)
	bulletRe  = regexp.MustCompile(`(?m)^[ \t]*[-*+][ \t]+`)
	orderedRe = regexp.MustCompile(`(?m)^[ \t]*\d+\.[ \t]+`)
	ruleRe    = regexp.MustCompile(`(?m)^[ \t]*(-{3,}|\*{3,}|_{3,})[ \t]*$\n?`)
	// Remaining tags, once the components are gone.
	tagRe        = regexp.MustCompile(`</?[A-Za-z][^>]*>`)
	inlineCodeRe = regexp.MustCompile("`([^`]*)`")
	// Emphasis reads as the words inside it. One pattern per marker, because
	// RE2 has no backreference to say "the same marker again".
	emphasisRes = []*regexp.Regexp{
		regexp.MustCompile(`\*\*\*([^*]+)\*\*\*`),
		regexp.MustCompile(`\*\*([^*]+)\*\*`),
		regexp.MustCompile(`\*([^*]+)\*`),
		regexp.MustCompile(`___([^_]+)___`),
		regexp.MustCompile(`__([^_]+)__`),
		regexp.MustCompile(`_([^_\s][^_]*)_`),
	}
	blankLinesRe = regexp.MustCompile(`\n{3,}`)
	spacesRe     = regexp.MustCompile(`[^\S\n]+`)
)

func narrationOf(source string) string {
	meta, body := splitFrontmatter(source)

	body = jsxBlockRe.ReplaceAllString(body, "\n\n")
	body = fencedCodeRe.ReplaceAllString(body, "\n\n")
	body = tableRe.ReplaceAllString(body, "")
	body = imageRe.ReplaceAllString(body, "")
	body = ruleRe.ReplaceAllString(body, "\n\n")
	body = linkRe.ReplaceAllString(body, "$1")
	body = headingRe.ReplaceAllString(body, "")
	body = quoteRe.ReplaceAllString(body, "")
	body = bulletRe.ReplaceAllString(body, "")
	body = orderedRe.ReplaceAllString(body, "")
	body = tagRe.ReplaceAllString(body, "")
	body = inlineCodeRe.ReplaceAllString(body, "$1")
	for _, re := range emphasisRes {
		body = re.ReplaceAllString(body, "$1")
	}

	// A blank line between parts is where Split prefers to cut, and cutting
	// between paragraphs is inaudible where cutting mid-sentence is not.
	parts := []string{meta["titre"], meta["chapeau"], body}
	var kept []string
	for _, p := range parts {
		if p = strings.TrimSpace(p); p != "" {
			kept = append(kept, p)
		}
	}
	return normalize(strings.Join(kept, "\n\n"))
}

// splitFrontmatter returns the frontmatter fields the narration needs and the
// body below them. It reads the two scalars it uses rather than the whole
// document: a YAML parser here would be a dependency for two strings.
func splitFrontmatter(source string) (map[string]string, string) {
	meta := map[string]string{}
	match := frontmatterRe.FindStringSubmatch(source)
	if match == nil {
		return meta, source
	}

	for _, line := range strings.Split(match[1], "\n") {
		key, value, found := strings.Cut(line, ":")
		if !found || strings.HasPrefix(strings.TrimSpace(key), "#") {
			continue
		}
		key = strings.TrimSpace(key)
		if key != "titre" && key != "chapeau" {
			continue
		}
		meta[key] = unquote(strings.TrimSpace(value))
	}
	return meta, source[len(match[0]):]
}

func unquote(s string) string {
	if len(s) >= 2 {
		if (s[0] == '"' && s[len(s)-1] == '"') || (s[0] == '\'' && s[len(s)-1] == '\'') {
			return s[1 : len(s)-1]
		}
	}
	return s
}

// normalize collapses the whitespace the stripping leaves behind, keeping
// paragraph breaks: they are the only whitespace a voice reads.
func normalize(s string) string {
	s = strings.ReplaceAll(s, "\r\n", "\n")
	s = spacesRe.ReplaceAllString(s, " ")
	var lines []string
	for _, line := range strings.Split(s, "\n") {
		lines = append(lines, strings.TrimSpace(line))
	}
	s = strings.Join(lines, "\n")
	s = blankLinesRe.ReplaceAllString(s, "\n\n")
	return strings.TrimSpace(s)
}
